needy/caches/directorycache.py

```diff
@@ -0,0 +1,55 @@
+import tarfile
+import os
+from datetime import datetime, timedelta
+
+from ..cache import Cache
+
+
+class DirectoryCache(Cache):
+    def __init__(self, configuration):
+        Cache.__init__(self, configuration)
+
+    @staticmethod
+    def type():
+        return 'directory'
+
+    def description(self):
+        return self.__path
+
+    def archive(self, source, key):
+        destination_path = self.__archive_path(key)
+        parent_path = os.path.dirname(destination_path)
+        if not os.path.exists(parent_path):
+            os.makedirs(parent_path)
+        if os.path.exists(destination_path):
+            os.remove(destination_path)
+        tar = tarfile.open(destination_path, 'w:gz')
+        tar.add(source, arcname='.')
+        tar.close()
+        self.__garbage_collect()
+
+    def unarchive(self, destination, key):
+        tar = tarfile.open(self.__archive_path(key), 'r:gz')
+        tar.extractall(path=destination)
+        tar.close()
+        os.utime(self.__archive_path(key), None)
+
+    def exists(self, key):
+        return os.path.exists(self.__archive_path(key))
+
+    def __garbage_collect(self, min_date=datetime.now()+timedelta(weeks=-2)):
+        for root, directories, filenames in os.walk(self.__archive_path()):
+            for filename in filenames:
+                path = os.path.join(root, filename)
+                try:
+                    mtime = os.path.getmtime(path)
+                except OSError:
+                    mtime = 0
+                last_modified_date = datetime.fromtimestamp(mtime)
+                if last_modified_date < min_date:
+                    logging.debug('Removing older cache item {}'.format(path))
+                    os.remove(path)
```

Comment:
In the interest of future extensibility, it would be really nice to put as much of the garbage collection logic in the superclass as possible.

You should probably garbage collect based on last used time, right?

Since a cache will potentially be shared across several projects, I think there are also atomicity concerns. It would definitely be possible for multiple instances to step on each other's toes here.